### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by feymartynov**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -123,40 +225,171 @@ impl FromStr for SharedGroup {
...
     Broadcast(&'a AccountId, &'a str, &'a str),
-    // <- request(app-from-any): agents/+/api/+/out/ACCOUNT_ID(ME)
-    // <- request(app-from-any): agents/+/api/VER/out/ACCOUNT_ID(ME)
-    // <- request(app-from-any): agents/AGENT_ID/api/+/out/ACCOUNT_ID(ME)
-    // <- request(app-from-any): agents/AGENT_ID/api/VER/out/ACCOUNT_ID(ME)
+    /// Receive a message from any single of [SharedGroup](struct.SharedGroup.html) agents.
+    ///
+    /// Typically used for receiving requests by services which may have multiple instances.
+    ///
+    /// [AccountId](struct.AccountId.html) is being supplied since we don't care which specific
```
Comment (manifest) [inline src/lib.rs]: the specified account identifier limits the subscription to its agents

File: src/lib.rs [hunk 2]
```diff
@@ -123,40 +225,171 @@ impl FromStr for SharedGroup {
...
+    /// | request  | one-to-one  | agents/`AGENT_ID`/api/`VER`/in/`MY_ACCOUNT_ID` |
+    /// | response | one-to_one  | agents/`AGENT_ID`/api/`VER`/in/`MY_ACCOUNT_ID` |
     Unicast(AgentId, String),
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
+/// Message source.
+///
+/// This is an abstraction over MQTT topic pattern to determine the subscription topic to receive
```
Comment (manifest) [inline src/lib.rs]: NOTE: The Source isn't intended for direct use.

File: src/lib.rs [hunk 3]
```diff
@@ -123,40 +225,171 @@ impl FromStr for SharedGroup {
...
     Unicast(Option<&'a AccountId>),
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
+/// Messages subscription builder.
 pub struct Subscription {}
 
 impl Subscription {
+    /// Builds an [EventSubscription](struct.EventSubscription) for
```
Comment (manifest) [inline src/lib.rs]: NOTE: The Subscription is intended for direct use.

File: src/lib.rs (hunk 2 above)
Comment (manifest) [reply src/lib.rs]: Maybe we should hide it from docs?

File: src/lib.rs [hunk 4]
```diff
@@ -123,40 +225,171 @@ impl FromStr for SharedGroup {
...
+/// * `MY_BROADCAST_URI`– current agent's API specific path to some resource divided by `/`.
+/// For example: `/rooms/ROOM_ID/events`. If you will want to change its structure in the future
+/// you must also bump `VER(ME)`.
+/// * `ACCOUNT_ID` – destination [AccountId](struct.AccountId) (no specific agent).
+/// * `AGENT_ID` – destination [AgentId](struct.AgentId).
+/// * `VER` – destination agent version.
 #[derive(Debug)]
 pub enum Destination {
-    // -> event(app-to-any): apps/ACCOUNT_ID(ME)/api/VER(ME)/BROADCAST_URI
+    /// Publish a message to each of the topic subscribers.
```
Comment (manifest) [inline src/lib.rs]: NOTE: The Destination isn't intended for direct use.

Comment (manifest) [reply src/lib.rs]: Maybe we should hide it from the docs?

File: src/lib.rs (hunk 3 above)
Comment (manifest) [reply src/lib.rs]: Maybe a good idea to specify which subscriptions are accessible for user / service agents.

File: src/lib.rs [hunk 5]
```diff
@@ -210,6 +524,19 @@ pub struct EventSubscription<'a> {
 }
 
 impl<'a> EventSubscription<'a> {
+    /// Builds an [EventSubscription](struct.EventSubscription).
```
Comment (manifest) [inline src/lib.rs]: NOTE: it's not intended for use by apps developers. Better to hide from the docs.

File: src/lib.rs [hunk 6]
```diff
@@ -220,6 +547,17 @@ pub struct RequestSubscription<'a> {
 }
 
 impl<'a> RequestSubscription<'a> {
+    /// Builds a [RequestSubscription](struct.RequestSubscription).
```
Comment (manifest) [inline src/lib.rs]: NOTE: it's not intended for use by apps developers. Better to hide from the docs.

File: src/lib.rs [hunk 7]
```diff
@@ -230,13 +568,30 @@ pub struct ResponseSubscription<'a> {
 }
 
 impl<'a> ResponseSubscription<'a> {
+    /// Builds a [ResponseSubscription](struct.ResponseSubscription).
```
Comment (manifest) [inline src/lib.rs]: NOTE: it's not intended for use by apps developers. Better to hide from the docs.

File: src/mqtt.rs [hunk 8]
```diff
@@ -179,11 +294,28 @@ impl Agent {
 
 ////////////////////////////////////////////////////////////////////////////////
 
+/// Connection mode of an agent that defines the level of privileges.
 #[derive(Debug, Clone)]
 pub enum ConnectionMode {
+    /// This mode locks the agent in his home topic allowing to publish and subscribe only to
```
Comment (manifest) [inline src/mqtt.rs]: Maybe explicitly state which message patterns are available for the each particular mode?

Comment (manifest) [inline src/mqtt.rs]: The agent in this mode may only publish to its "home" topic: "agents/..."

File: src/mqtt.rs [hunk 9]
```diff
@@ -323,6 +455,20 @@ impl Addressable for ConnectionProperties {
...
 ////////////////////////////////////////////////////////////////////////////////
 
+/// Timing properties that persist through a message chain.
+///
+/// See [ShortTermTimingProperties](ShortTermTimingProperties.html) for more explanation
+/// on timings.
+///
+/// There are two kinds of properties: regular an cumulative.
+/// Regular properties just get proxied without change to the next message in the chain.
+/// Cumulative properties sum corresponding values from
```
Comment (manifest) [inline src/mqtt.rs]: sum -> accumulate?

File: src/mqtt.rs [hunk 10]
```diff
@@ -635,6 +934,27 @@ impl IncomingRequestProperties {
...
 
+    /// Builds [OutgoingRequestProperties](struct.OutgoingRequestProperties.html) based on the
+    /// [IncomingRequestProperties](struct.IncomingRequestProperties.html).
+    ///
+    /// Use it to send a request to another service while handling a request.
+    ///
+    /// # Arguments
+    ///
+    /// * `method` – request method.
+    /// * `response_topic` – a topic to send the response to the request to.
```
Comment (manifest) [inline src/mqtt.rs]: topic for reponse

File: src/mqtt.rs [hunk 11]
```diff
@@ -635,6 +934,27 @@ impl IncomingRequestProperties {
...
+    /// * `correlation_data` – any string to correlate request with the upcoming response.
```
Comment (manifest) [inline src/mqtt.rs]: any string to correlate request with response (response is outgoing)

File: src/mqtt.rs [hunk 12]
```diff
@@ -774,6 +1112,26 @@ where
...
 
 impl<T> IncomingRequest<T> {
+    /// Builds [OutgoingResponse](OutgoingResponse.html) based on
+    /// the [IncomingRequest](IncomingRequest.html).
+    ///
+    /// Use it to response on a request.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` – serializable response payload.
```
Comment (manifest) [inline src/mqtt.rs]: maybe just "JSON payload"?

File: src/mqtt.rs [hunk 13]
```diff
@@ -901,6 +1307,34 @@ pub struct OutgoingResponseProperties {
...
+    ///
+    /// Generally you shouldn't use this function and consider using
+    /// [IncomingRequestProperties::to_response](struct.IncomingRequestProperties.html#method.to_respone)
+    /// because all outgoing responses are related to an incoming request to respond to.
+    /// However if you need to customize the response creation you may want to call this constructor
+    /// directly.
+    ///
+    /// # Arguments
+    ///
+    /// * `status` – response status.
```
Comment (manifest) [inline src/mqtt.rs]: HTTP-compatible status code

File: src/mqtt.rs [hunk 14]
```diff
@@ -901,6 +1307,34 @@ pub struct OutgoingResponseProperties {
...
+    /// * `correlation_data` – the correlation string from the request so that the agent that made
```
Comment (manifest) [inline src/mqtt.rs]: A correlation between request and response. It has meaning to the sender of the request message and receiver of the response message.

File: src/lib.rs [hunk 15]
```diff
@@ -77,13 +148,28 @@ impl Addressable for AgentId {
 
 ////////////////////////////////////////////////////////////////////////////////
 
+/// A group of agents which shares a subscription.
```
[the remark this answers is not in the document]
Comment (feymartynov) [reply src/lib.rs]: > or just `the shared subscription`

I think the current variant is more clear for those who faces the term for the first time.

File: src/mqtt.rs (hunk 12 above)
Comment (feymartynov) [reply src/mqtt.rs]: I wanted to emphasize that it's an object that implements `Serializable` trait, not a string with dumped JSON for example.

**Pull request merged**